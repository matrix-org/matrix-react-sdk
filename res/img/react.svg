<svg xmlns="http://www.w3.org/2000/svg" width="18" height="17" viewBox="0 0 18 17">
    <g fill="none" fill-rule="evenodd">
        <path fill="#2E2F32" fill-rule="nonzero" d="M13.6 14.5a7.875 7.875 0 1 1 2.183-3h-.818a7.125 7.125 0 1 0-2.62 3H13.6z"/>
        <g stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75">
            <path d="M11.5 13h6M14.5 10v6"/>
        </g>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75" d="M5.5 10s1.125 1.5 3 1.5 3-1.5 3-1.5"/>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M6.25 6.25h.008M10.75 6.25h.008"/>
    </g>
</svg>
