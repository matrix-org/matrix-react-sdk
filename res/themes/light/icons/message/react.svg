<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><path d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 464c-119.1 0-216-96.9-216-216S128.9 40 248 40s216 96.9 216 216-96.9 216-216 216zm90.2-146.2C315.8 352.6 282.9 368 248 368s-67.8-15.4-90.2-42.2c-5.7-6.8-15.8-7.7-22.5-2-6.8 5.7-7.7 15.7-2 22.5C161.7 380.4 203.6 400 248 400s86.3-19.6 114.8-53.8c5.7-6.8 4.8-16.9-2-22.5-6.8-5.6-16.9-4.7-22.6 2.1zM168 240c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32zm160 0c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32z"/></svg>