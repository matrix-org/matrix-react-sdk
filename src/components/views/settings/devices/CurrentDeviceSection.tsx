/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, { useState } from "react";
import { LocalNotificationSettings } from "matrix-js-sdk/src/matrix";

import { _t } from "../../../../languageHandler";
import Spinner from "../../elements/Spinner";
import SettingsSubsection from "../shared/SettingsSubsection";
import { SettingsSubsectionHeading } from "../shared/SettingsSubsectionHeading";
import DeviceDetails from "./DeviceDetails";
import { DeviceExpandDetailsButton } from "./DeviceExpandDetailsButton";
import DeviceTile from "./DeviceTile";
import { DeviceVerificationStatusCard } from "./DeviceVerificationStatusCard";
import { ExtendedDevice } from "./types";
import { KebabContextMenu } from "../../context_menus/KebabContextMenu";
import { IconizedContextMenuOption } from "../../context_menus/IconizedContextMenu";

interface Props {
    device?: ExtendedDevice;
    isLoading: boolean;
    isSigningOut: boolean;
    localNotificationSettings?: LocalNotificationSettings;
    // number of other sessions the user has
    // excludes current session
    otherSessionsCount: number;
    setPushNotifications: (deviceId: string, enabled: boolean) => Promise<void>;
    onVerifyCurrentDevice: () => void;
    onSignOutCurrentDevice: () => void;
    signOutAllOtherSessions?: () => void;
    saveDeviceName: (deviceName: string) => Promise<void>;
}

type CurrentDeviceSectionHeadingProps = Pick<
    Props,
    "onSignOutCurrentDevice" | "signOutAllOtherSessions" | "otherSessionsCount"
> & {
    disabled?: boolean;
};

const CurrentDeviceSectionHeading: React.FC<CurrentDeviceSectionHeadingProps> = ({
    onSignOutCurrentDevice,
    signOutAllOtherSessions,
    otherSessionsCount,
    disabled,
}) => {
    const menuOptions = [
        <IconizedContextMenuOption
            key="sign-out"
            label={_t("action|sign_out")}
            onClick={onSignOutCurrentDevice}
            isDestructive
        />,
        ...(signOutAllOtherSessions
            ? [
                  <IconizedContextMenuOption
                      key="sign-out-all-others"
                      label={_t("settings|sessions|sign_out_all_other_sessions", { otherSessionsCount })}
                      onClick={signOutAllOtherSessions}
                      isDestructive
                  />,
              ]
            : []),
    ];
    return (
        <SettingsSubsectionHeading heading={_t("settings|sessions|current_session")}>
            <KebabContextMenu
                disabled={disabled}
                title={_t("common|options")}
                options={menuOptions}
                data-testid="current-session-menu"
            />
        </SettingsSubsectionHeading>
    );
};

const CurrentDeviceSection: React.FC<Props> = ({
    device,
    isLoading,
    isSigningOut,
    localNotificationSettings,
    otherSessionsCount,
    setPushNotifications,
    onVerifyCurrentDevice,
    onSignOutCurrentDevice,
    signOutAllOtherSessions,
    saveDeviceName,
}) => {
    const [isExpanded, setIsExpanded] = useState(false);

    return (
        <SettingsSubsection
            data-testid="current-session-section"
            heading={
                <CurrentDeviceSectionHeading
                    onSignOutCurrentDevice={onSignOutCurrentDevice}
                    signOutAllOtherSessions={signOutAllOtherSessions}
                    otherSessionsCount={otherSessionsCount}
                    disabled={isLoading || !device || isSigningOut}
                />
            }
        >
            {/* only show big spinner on first load */}
            {isLoading && !device && <Spinner />}
            {!!device && (
                <>
                    <DeviceTile device={device} onClick={() => setIsExpanded(!isExpanded)}>
                        <DeviceExpandDetailsButton
                            data-testid="current-session-toggle-details"
                            isExpanded={isExpanded}
                            onClick={() => setIsExpanded(!isExpanded)}
                        />
                    </DeviceTile>
                    {isExpanded ? (
                        <DeviceDetails
                            device={device}
                            localNotificationSettings={localNotificationSettings}
                            setPushNotifications={setPushNotifications}
                            isSigningOut={isSigningOut}
                            onVerifyDevice={onVerifyCurrentDevice}
                            onSignOutDevice={onSignOutCurrentDevice}
                            saveDeviceName={saveDeviceName}
                            className="mx_CurrentDeviceSection_deviceDetails"
                        />
                    ) : (
                        <>
                            <br />
                            <DeviceVerificationStatusCard
                                device={device}
                                onVerifyDevice={onVerifyCurrentDevice}
                                isCurrentDevice
                            />
                        </>
                    )}
                </>
            )}
        </SettingsSubsection>
    );
};

export default CurrentDeviceSection;
