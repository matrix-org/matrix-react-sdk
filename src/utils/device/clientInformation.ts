/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { MatrixClient } from "matrix-js-sdk/src/matrix";

import BasePlatform from "../../BasePlatform";
import { IConfigOptions } from "../../IConfigOptions";
import { DeepReadonly } from "../../@types/common";

export type DeviceClientInformation = {
    name?: string;
    version?: string;
    url?: string;
};

const formatUrl = (): string | undefined => {
    // don't record url for electron clients
    if (window.electron) {
        return undefined;
    }

    // strip query-string and fragment from uri
    const url = new URL(window.location.href);

    return [
        url.host,
        url.pathname.replace(/\/$/, ""), // Remove trailing slash if present
    ].join("");
};

const clientInformationEventPrefix = "io.element.matrix_client_information.";
export const getClientInformationEventType = (deviceId: string): string => `${clientInformationEventPrefix}${deviceId}`;

/**
 * Record extra client information for the current device
 * https://github.com/vector-im/element-meta/blob/develop/spec/matrix_client_information.md
 */
export const recordClientInformation = async (
    matrixClient: MatrixClient,
    sdkConfig: DeepReadonly<IConfigOptions>,
    platform?: BasePlatform,
): Promise<void> => {
    const deviceId = matrixClient.getDeviceId()!;
    const { brand } = sdkConfig;
    const version = await platform?.getAppVersion();
    const type = getClientInformationEventType(deviceId);
    const url = formatUrl();

    await matrixClient.setAccountData(type, {
        name: brand,
        version,
        url,
    });
};

/**
 * Remove client information events for devices that no longer exist
 * @param validDeviceIds - ids of current devices,
 *                      client information for devices NOT in this list will be removed
 */
export const pruneClientInformation = (validDeviceIds: string[], matrixClient: MatrixClient): void => {
    Array.from(matrixClient.store.accountData.values()).forEach((event) => {
        if (!event.getType().startsWith(clientInformationEventPrefix)) {
            return;
        }
        const [, deviceId] = event.getType().split(clientInformationEventPrefix);
        if (deviceId && !validDeviceIds.includes(deviceId)) {
            matrixClient.deleteAccountData(event.getType());
        }
    });
};

/**
 * Remove extra client information for current device
 */
export const removeClientInformation = async (matrixClient: MatrixClient): Promise<void> => {
    const deviceId = matrixClient.getDeviceId()!;
    const type = getClientInformationEventType(deviceId);
    const clientInformation = getDeviceClientInformation(matrixClient, deviceId);

    // if a non-empty client info event exists, remove it
    if (clientInformation.name || clientInformation.version || clientInformation.url) {
        await matrixClient.deleteAccountData(type);
    }
};

const sanitizeContentString = (value: unknown): string | undefined =>
    value && typeof value === "string" ? value : undefined;

export const getDeviceClientInformation = (matrixClient: MatrixClient, deviceId: string): DeviceClientInformation => {
    const event = matrixClient.getAccountData(getClientInformationEventType(deviceId));

    if (!event) {
        return {};
    }

    const { name, version, url } = event.getContent();

    return {
        name: sanitizeContentString(name),
        version: sanitizeContentString(version),
        url: sanitizeContentString(url),
    };
};
