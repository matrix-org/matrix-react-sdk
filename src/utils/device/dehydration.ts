/*
Copyright 2024 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { logger } from "matrix-js-sdk/src/logger";
import { MatrixClientPeg } from "../../MatrixClientPeg";

// the interval between creating dehydrated devices
const DEHYDRATION_INTERVAL = 7*24*60*60*1000;

// check if device dehydration is enabled
export async function deviceDehydrationEnabled(): Promise<boolean> {
    const crypto = MatrixClientPeg.safeGet().getCrypto();
    if (!(await crypto.isDehydrationSupported())) {
        return false;
    }
    if (await crypto.isDehydrationKeyStored()) {
        return true;
    }
    return true;
    const wellknown = await MatrixClientPeg.safeGet().waitForClientWellKnown();
    return !!wellknown?.["org.matrix.msc3814"];
}

// if dehydration is enabled, rehydrate a device (if available) and create
// a new dehydrated device
export async function initializeDehydration(reset?: boolean): Promise<void> {
    const crypto = MatrixClientPeg.safeGet().getCrypto();
    if (crypto && await deviceDehydrationEnabled()) {
        logger.log("Device dehydration enabled");
        if (reset) {
            await crypto.resetDehydrationKey();
        } else {
            try {
                await crypto.rehydrateDeviceIfAvailable();
            } catch (e) {
                logger.error("Error rehydrating device:", e);
            }
        }
        await crypto.scheduleDeviceDehydration(DEHYDRATION_INTERVAL);
    }
}
