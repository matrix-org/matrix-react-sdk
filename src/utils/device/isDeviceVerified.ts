/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { MatrixClient } from "matrix-js-sdk/src/matrix";

/**
 * Check if one of our own devices is verified via cross signing
 *
 * @param client - reference to the MatrixClient
 * @param deviceId - ID of the device to be checked
 *
 * @returns `null` if the device is unknown or has not published encryption keys; otherwise a boolean
 *    indicating whether the device has been cross-signed by a cross-signing key we trust.
 */
export const isDeviceVerified = async (client: MatrixClient, deviceId: string): Promise<boolean | null> => {
    const trustLevel = await client.getCrypto()?.getDeviceVerificationStatus(client.getSafeUserId(), deviceId);
    if (!trustLevel) {
        // either no crypto, or an unknown/no-e2e device
        return null;
    }
    return trustLevel.crossSigningVerified;
};
