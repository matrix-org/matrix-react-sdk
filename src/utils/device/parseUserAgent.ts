/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import UAParser from "ua-parser-js";

export enum DeviceType {
    Desktop = "Desktop",
    Mobile = "Mobile",
    Web = "Web",
    Unknown = "Unknown",
}
export type ExtendedDeviceInformation = {
    deviceType: DeviceType;
    // eg Google Pixel 6
    deviceModel?: string;
    // eg Android 11
    deviceOperatingSystem?: string;
    // eg Firefox 1.1.0
    client?: string;
};

// Element/1.8.21 (iPhone XS Max; iOS 15.2; Scale/3.00)
const IOS_KEYWORD = "; iOS ";
const BROWSER_KEYWORD = "Mozilla/";

const getDeviceType = (
    userAgent: string,
    device: UAParser.IDevice,
    browser: UAParser.IBrowser,
    operatingSystem: UAParser.IOS,
): DeviceType => {
    if (browser.name === "Electron") {
        return DeviceType.Desktop;
    }
    if (!!browser.name) {
        return DeviceType.Web;
    }
    if (device.type === "mobile" || operatingSystem.name?.includes("Android") || userAgent.indexOf(IOS_KEYWORD) > -1) {
        return DeviceType.Mobile;
    }
    return DeviceType.Unknown;
};

interface CustomValues {
    customDeviceModel?: string;
    customDeviceOS?: string;
}
/**
 * Some mobile model and OS strings are not recognised
 * by the UA parsing library
 * check they exist by hand
 */
const checkForCustomValues = (userAgent: string): CustomValues => {
    if (userAgent.includes(BROWSER_KEYWORD)) {
        return {};
    }

    const mightHaveDevice = userAgent.includes("(");
    if (!mightHaveDevice) {
        return {};
    }
    const deviceInfoSegments = userAgent.substring(userAgent.indexOf("(") + 1).split("; ");
    const customDeviceModel = deviceInfoSegments[0] || undefined;
    const customDeviceOS = deviceInfoSegments[1] || undefined;
    return { customDeviceModel, customDeviceOS };
};

const concatenateNameAndVersion = (name?: string, version?: string): string | undefined =>
    name && [name, version].filter(Boolean).join(" ");

export const parseUserAgent = (userAgent?: string): ExtendedDeviceInformation => {
    if (!userAgent) {
        return {
            deviceType: DeviceType.Unknown,
        };
    }

    const parser = new UAParser(userAgent);

    const browser = parser.getBrowser();
    const device = parser.getDevice();
    const operatingSystem = parser.getOS();

    const deviceType = getDeviceType(userAgent, device, browser, operatingSystem);

    // OSX versions are frozen at 10.15.17 in UA strings https://chromestatus.com/feature/5452592194781184
    // ignore OS version in browser based sessions
    const shouldIgnoreOSVersion = deviceType === DeviceType.Web || deviceType === DeviceType.Desktop;
    const deviceOperatingSystem = concatenateNameAndVersion(
        operatingSystem.name,
        shouldIgnoreOSVersion ? undefined : operatingSystem.version,
    );
    const deviceModel = concatenateNameAndVersion(device.vendor, device.model);
    const client = concatenateNameAndVersion(browser.name, browser.version);

    // only try to parse custom model and OS when device type is known
    const { customDeviceModel, customDeviceOS } =
        deviceType !== DeviceType.Unknown ? checkForCustomValues(userAgent) : ({} as CustomValues);

    return {
        deviceType,
        deviceModel: deviceModel || customDeviceModel,
        deviceOperatingSystem: deviceOperatingSystem || customDeviceOS,
        client,
    };
};
