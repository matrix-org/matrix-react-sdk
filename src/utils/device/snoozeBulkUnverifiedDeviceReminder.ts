/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { logger } from "matrix-js-sdk/src/logger";

const SNOOZE_KEY = "mx_snooze_bulk_unverified_device_nag";
// one week
const snoozePeriod = 1000 * 60 * 60 * 24 * 7;
export const snoozeBulkUnverifiedDeviceReminder = (): void => {
    try {
        localStorage.setItem(SNOOZE_KEY, String(Date.now()));
    } catch (error) {
        logger.error("Failed to persist bulk unverified device nag snooze", error);
    }
};

export const isBulkUnverifiedDeviceReminderSnoozed = (): boolean => {
    try {
        const snoozedTimestamp = localStorage.getItem(SNOOZE_KEY);

        const parsedTimestamp = Number.parseInt(snoozedTimestamp || "", 10);

        return Number.isInteger(parsedTimestamp) && parsedTimestamp + snoozePeriod > Date.now();
    } catch (error) {
        return false;
    }
};
