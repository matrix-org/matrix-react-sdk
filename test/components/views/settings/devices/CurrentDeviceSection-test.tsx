/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import { act, fireEvent, render } from "@testing-library/react";

import CurrentDeviceSection from "../../../../../src/components/views/settings/devices/CurrentDeviceSection";
import { DeviceType } from "../../../../../src/utils/device/parseUserAgent";

describe("<CurrentDeviceSection />", () => {
    const deviceId = "alices_device";

    const alicesVerifiedDevice = {
        device_id: deviceId,
        isVerified: false,
        deviceType: DeviceType.Unknown,
    };
    const alicesUnverifiedDevice = {
        device_id: deviceId,
        isVerified: false,
        deviceType: DeviceType.Unknown,
    };

    const defaultProps = {
        device: alicesVerifiedDevice,
        onVerifyCurrentDevice: jest.fn(),
        onSignOutCurrentDevice: jest.fn(),
        saveDeviceName: jest.fn(),
        isLoading: false,
        isSigningOut: false,
        otherSessionsCount: 1,
        setPushNotifications: jest.fn(),
    };

    const getComponent = (props = {}): React.ReactElement => <CurrentDeviceSection {...defaultProps} {...props} />;

    it("renders spinner while device is loading", () => {
        const { container } = render(getComponent({ device: undefined, isLoading: true }));
        expect(container.getElementsByClassName("mx_Spinner").length).toBeTruthy();
    });

    it("handles when device is falsy", async () => {
        const { container } = render(getComponent({ device: undefined }));
        expect(container).toMatchSnapshot();
    });

    it("renders device and correct security card when device is verified", () => {
        const { container } = render(getComponent());
        expect(container).toMatchSnapshot();
    });

    it("renders device and correct security card when device is unverified", () => {
        const { container } = render(getComponent({ device: alicesUnverifiedDevice }));
        expect(container).toMatchSnapshot();
    });

    it("displays device details on main tile click", () => {
        const { getByTestId, container } = render(getComponent({ device: alicesUnverifiedDevice }));

        act(() => {
            fireEvent.click(getByTestId(`device-tile-${alicesUnverifiedDevice.device_id}`));
        });

        expect(container.getElementsByClassName("mx_DeviceDetails").length).toBeTruthy();

        act(() => {
            fireEvent.click(getByTestId(`device-tile-${alicesUnverifiedDevice.device_id}`));
        });

        // device details are hidden
        expect(container.getElementsByClassName("mx_DeviceDetails").length).toBeFalsy();
    });

    it("displays device details on toggle click", () => {
        const { container, getByTestId } = render(getComponent({ device: alicesUnverifiedDevice }));

        act(() => {
            fireEvent.click(getByTestId("current-session-toggle-details"));
        });

        expect(container.getElementsByClassName("mx_DeviceDetails")).toMatchSnapshot();

        act(() => {
            fireEvent.click(getByTestId("current-session-toggle-details"));
        });

        // device details are hidden
        expect(container.getElementsByClassName("mx_DeviceDetails").length).toBeFalsy();
    });
});
