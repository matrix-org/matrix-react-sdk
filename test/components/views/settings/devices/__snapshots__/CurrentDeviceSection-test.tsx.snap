// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CurrentDeviceSection /> displays device details on toggle click 1`] = `
HTMLCollection [
  <div
    class="mx_DeviceDetails"
    data-testid="device-detail-alices_device"
  >
    <section
      class="mx_DeviceDetails_section"
    >
      <h3
        class="mx_Heading_h3"
      >
        alices_device
      </h3>
      <div
        class="mx_DeviceSecurityCard"
      >
        <div
          class="mx_DeviceSecurityCard_icon Unverified"
        >
          <div
            height="16"
            width="16"
          />
        </div>
        <div
          class="mx_DeviceSecurityCard_content"
        >
          <p
            class="mx_DeviceSecurityCard_heading"
          >
            Unverified session
          </p>
          <p
            class="mx_DeviceSecurityCard_description"
          >
            Verify or sign out from this session for best security and reliability.
          </p>
        </div>
      </div>
    </section>
    <section
      class="mx_DeviceDetails_section"
    >
      <p
        class="mx_DeviceDetails_sectionHeading"
      >
        Session details
      </p>
      <table
        class="mxDeviceDetails_metadataTable"
      >
        <tbody>
          <tr>
            <td
              class="mxDeviceDetails_metadataLabel"
            >
              Session ID
            </td>
            <td
              class="mxDeviceDetails_metadataValue"
            >
              alices_device
            </td>
          </tr>
          <tr>
            <td
              class="mxDeviceDetails_metadataLabel"
            >
              Last activity
            </td>
            <td
              class="mxDeviceDetails_metadataValue"
            />
          </tr>
        </tbody>
      </table>
      <table
        class="mxDeviceDetails_metadataTable"
      >
        <thead>
          <tr>
            <th>
              Device
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="mxDeviceDetails_metadataLabel"
            >
              IP address
            </td>
            <td
              class="mxDeviceDetails_metadataValue"
            />
          </tr>
        </tbody>
      </table>
    </section>
  </div>,
]
`;

exports[`<CurrentDeviceSection /> handles when device is falsy 1`] = `
<div>
  <div
    class="mx_SettingsSubsection"
    data-testid="current-session-section"
  >
    <h3
      class="mx_Heading_h3 mx_SettingsSubsection_heading"
    >
      Current session
    </h3>
    <div
      class="mx_SettingsSubsection_content"
    />
  </div>
</div>
`;

exports[`<CurrentDeviceSection /> renders device and correct security card when device is unverified 1`] = `
<div>
  <div
    class="mx_SettingsSubsection"
    data-testid="current-session-section"
  >
    <h3
      class="mx_Heading_h3 mx_SettingsSubsection_heading"
    >
      Current session
    </h3>
    <div
      class="mx_SettingsSubsection_content"
    >
      <div
        class="mx_DeviceTile"
        data-testid="device-tile-alices_device"
      >
        <div
          class="mx_DeviceTile_info"
        >
          <h4
            class="mx_Heading_h4"
          >
            alices_device
          </h4>
          <div
            class="mx_DeviceTile_metadata"
          >
            <span
              data-testid="device-metadata-isVerified"
            >
              Unverified
            </span>
             路 
             路 
          </div>
        </div>
        <div
          class="mx_DeviceTile_actions"
        >
          <div
            aria-label="Toggle device details"
            class="mx_AccessibleButton mx_DeviceExpandDetailsButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_icon"
            data-testid="current-session-toggle-details"
            role="button"
            tabindex="0"
          >
            <div
              class="mx_DeviceExpandDetailsButton_icon"
            />
          </div>
        </div>
      </div>
      <br />
      <div
        class="mx_DeviceSecurityCard"
      >
        <div
          class="mx_DeviceSecurityCard_icon Unverified"
        >
          <div
            height="16"
            width="16"
          />
        </div>
        <div
          class="mx_DeviceSecurityCard_content"
        >
          <p
            class="mx_DeviceSecurityCard_heading"
          >
            Unverified session
          </p>
          <p
            class="mx_DeviceSecurityCard_description"
          >
            Verify or sign out from this session for best security and reliability.
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<CurrentDeviceSection /> renders device and correct security card when device is verified 1`] = `
<div>
  <div
    class="mx_SettingsSubsection"
    data-testid="current-session-section"
  >
    <h3
      class="mx_Heading_h3 mx_SettingsSubsection_heading"
    >
      Current session
    </h3>
    <div
      class="mx_SettingsSubsection_content"
    >
      <div
        class="mx_DeviceTile"
        data-testid="device-tile-alices_device"
      >
        <div
          class="mx_DeviceTile_info"
        >
          <h4
            class="mx_Heading_h4"
          >
            alices_device
          </h4>
          <div
            class="mx_DeviceTile_metadata"
          >
            <span
              data-testid="device-metadata-isVerified"
            >
              Unverified
            </span>
             路 
             路 
          </div>
        </div>
        <div
          class="mx_DeviceTile_actions"
        >
          <div
            aria-label="Toggle device details"
            class="mx_AccessibleButton mx_DeviceExpandDetailsButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_icon"
            data-testid="current-session-toggle-details"
            role="button"
            tabindex="0"
          >
            <div
              class="mx_DeviceExpandDetailsButton_icon"
            />
          </div>
        </div>
      </div>
      <br />
      <div
        class="mx_DeviceSecurityCard"
      >
        <div
          class="mx_DeviceSecurityCard_icon Unverified"
        >
          <div
            height="16"
            width="16"
          />
        </div>
        <div
          class="mx_DeviceSecurityCard_content"
        >
          <p
            class="mx_DeviceSecurityCard_heading"
          >
            Unverified session
          </p>
          <p
            class="mx_DeviceSecurityCard_description"
          >
            Verify or sign out from this session for best security and reliability.
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
